### ISSUE TITLE
Planned and unplanned EVA activities

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#511]
**Is your feature request related to a problem? Please describe.**
- Currently, there is no EVA planning. 
- EVA activities are driven by tasks at hand.

**Describe the solution you'd like**
- A settlement should have both planned EVA activities and unplanned EVA activities. 
- For planned activities, the settlement has a schedule of when and who and what will be accomplished and thus plan carefully prior to carrying them out. 
- At the minimum, it should be planned one day ahead.
- EVA repairs will be considered unplanned activities.

**Comment by bevans2000** [comment]
I uncovered an interesting problem recently with Building Airlocks. The Operator seems to get stuck entering the airlock because it is full BUT they are already down as an occupant. Not sure (and didn't investigate) how it happened. But I did change the  method ```EnterAirlock.enterAirlock``` as an experiment to
```
		if (!airlock.isOuterDoorLocked()) {
			boolean alreadyPresent = airlock.inAirlock(person);
			if (alreadyPresent || (!airlock.isChamberFull() && airlock.hasSpace())) {

				if (!alreadyPresent) {
					canProceed = airlock.enterAirlock(person, id, false);
				}
                                else {
                                ......
```
It seemed to fix the problem.

**Comment by mokun** [comment]
Currently, the definition of occupant isn't as clear as it should be.

If a person is in zone 2, 3, or 4, then he/she is considered an occupant.

Zone 3 is the chamber area where a person will don and doff an EVA suit.

One may say zone 2 and 4 as the stopping area inside the airlock. They may have just managed to come through the inner or outer doors. They may also have just done with donning or doffing the suit and waiting for the inner or outer door to open again.

**Comment by mokun** [comment]
To address the issue in https://github.com/mars-sim/mars-sim/commit/a18a2c849c325d1eef7d6661dde82c0d1c981bbf#commitcomment-83710759, 
use the concept of airlock mode in https://github.com/mars-sim/mars-sim/commit/6dd980692b6694f09d541ee6fa6a3652c4df6b62 .

**Comment by mokun** [comment]
@bevans2000 

While we've uncovered most if not all bugs regarding EVA, it is still somewhat disorganized for settlers in wanting to do an EVA. 

There are times when a settler who wants to do an EVA is forced to walk way from airlock due to various reasons such as fitness, others have gone too far ahead, no EVA suits, etc.

When it happens, other it tends to walk away with EVA suit and without garment and bottle, which I'll drop in a fix in a moment to brings this to a full closure.

Anyway, we still don't have *mass* EVA that are scheduled ahead of time.

**Comment by mokun** [comment]
@bevans2000  @Urwumpe 

What does it take to implement a well planned EVA schedule ?

Logistically speaking, I suppose we just need to gather all 4 settlers at one place. 

When they show up, we first qualify each of them, then we send them all to an available airlock.

They will go through egress steps at the same time, which will guarantee no one getting stuck, etc.

Not getting sync-up inside the airlock has been the biggest issue when it comes to opening and closing inner and outer doors, pressuring and depressing airlock, etc.

On top of that, we allow people to ingress and egress at the same time (see [#637](https://github.com/mars-sim/mars-sim/issues/637), for the sake of saving precious time when an airlock needs to cycle the air to pressurize and depressurize.

I'd really like to take care of this once and for all and move on.

**Comment by mokun** [comment]
@bevans2000 

I'm about to enable planned EVA activities.

All EVA activities should have been planned ahead of time.

So it actually makes sense for anyone to get the approval first.

Question: if we are to give authority or permission to commander (and player) to pre-approve for EVA, where do you see as the best way to programmatically put a block to ***stop*** normal EVA activities for now ? 

I suppose in future, we can implement the idea of settlement wide policies.

If a person wants to do something that's prohibited by under a certain police, he can even file a waiver to get an exception

**Comment by mokun** [comment]
@bevans2000 @Urwumpe 

I've created a rudimentary `ScheduleManager` and `Appointment` class a few commits ago and will put it to test.

**Comment by mokun** [comment]
I found it handy to use Settlement's two methods as follows : 

```java
	/**
	 * Sets the process override flag.
	 *
	 * @param type Name of process type
	 * @param override override for processes.
	 */
	public void setProcessOverride(OverrideType type, boolean override) {
		if (override) {
			logger.log(this, Level.CONFIG, 0L, "Player enables the override on '" + type.toString() + "'.");
			this.processOverrides.add(type);
		}
		else {
			logger.log(this, Level.CONFIG, 0L, "Player disables the override on '" + type.toString() + "'.");
			this.processOverrides.remove(type);
		}
	}

	/**
	 * Gets the process override flag.
	 *
	 * @param type Name of process type
	 * @return Is this override flag set
	 */
	public boolean getProcessOverride(OverrideType type) {
		return this.processOverrides.contains(type);
	}
```

I'll probably add a new OverrideType called DIGLOCAL to turn off DigLocal tasks unless it's a pre-planned task.

**Comment by mokun** [comment]
@bevans2000 @Urwumpe 

Can you see how we can improve on this ? 

Digging local regolith/ice seems to me a task for resource (and may be logistic) specialists to worry about. That's why I gave them that task. 

But this can go beyond planning for just EVA.

What was just committed was to serve as a springboard for what we can potentially do in terms of setting up a daily schedule for each role of a division.

There are still tons of things to iron out.

Any idea what to do next ?

**Comment by bevans2000** [comment]
I get the idea but we need to make it generic so we can have scheduled events for a Person. I wonder is we can just put these events into the Settlement future events queued such that we don't need a ScheduleManager for each Person which may become an overhead.

**Comment by mokun** [comment]
> these events into the Settlement future events queued such that we don't need a ScheduleManager for each Person which may become an overhead.

Yes programmatically we can have one big settlement handling the work schedule for everyone. 

![image](https://github.com/mars-sim/mars-sim/assets/1168584/ac5312ea-699f-465d-98bc-cb008987d30f)

Because I can see we already have a few implementation of future events for use in Transportable/Resupply, Shift Rotation, GoodsManager/Trading, etc., I'm seeing we could have a few implementation of this when it comes to scheduling tasks. 

The sandbox mode has a lot of freedom and is good for experimentation but it lacks purpose. 

I feel like giving each settlement some more definite goals is good. 

May be making players paying more attention to how a command structure or government is supposed to function in a healthy way is an interesting one to model.

So, other than fixing EVA with a schedule,  the very first thing I want to see is in having a schedule is giving each division a standard schedule. 

And you know each base already has a command hierarchy in place with commander and sub-commander and up to 8 divisions with chiefs and specialists, why don't we further design how each division is supposed to serve and function in mars-sim ? 

In future, I do like to give back each person a good degree of freedom in deciding what he's planning to do for himself personally. 

It's like in real world, the job we have imposed a schedule of its own upon our lives. We also have our own things we want to do but also what our family needs us us to do it, etc., each having its own rationale and purpose.

Anyway, may be it's not a bad or wasteful thing to depict how each settler in mars-sim as a *FREE AGENT* that has the autonomy in creating his own schedule and objectives.

But collectively speaking, the whole group of people will exhibit

**Comment by bevans2000** [comment]
Interesting idea. We need a bit of thought on this but I think we need a new Task that can be created when a Person has to wait for others. This would be a sub Task and shared by many Persons.

**Comment by mokun** [comment]
@Urwumpe @bevans2000 

We've been imposing a fitness standard in a lot of EVA related tasks since a year ago. 

This morning, it dawns on me that this isn't exactly a good idea to impose the SAME fitness standard on everyone.

Some are better in tolerating high physical activities than others. 

Therefore, we've been witnessing some settlers are able to carry more weight outdoor and have a longer duration of time longer than others.

So, it's better off we let everyone to decide if they need to rest or not, based upon their physical makeup.

This way, we should have less log statements complaining about not being able to fully finish the airlock activities to go out, etc.

In light of that, I'll add the use of the muscle tolerance param or something similar to modify those fitness related methods in `PhysicalCondition`.

The point is that unless it's a safety standard to impose, we should make outdoor activities as ***organic*** as possible.

**Comment by bevans2000** [comment]
Ok, the ExperienceImpact of the Task has a physical effort rating for the Task. So we should use that as well because it's an assessment of how physical a Task is.

**Comment by mokun** [comment]
> ExperienceImpact of the Task has a physical effort rating for the Task. 

So specifically in what way will it impact or improve the EVA op modeling ?

**Comment by bevans2000** [comment]
> > ExperienceImpact of the Task has a physical effort rating for the Task. 
> 
> So specifically in what way will it impact or improve the EVA op modeling ? 

It could be used to guage the fitness require to perform a Task

**Comment by mokun** [comment]
@bevans2000 

How do we use `ScheduledEventManager` exclusively for scheduling EVA tasks ahead of time in order to avoid airlock usage conflict ?

If each airlock has a usage schedule, would this problem be at least half-way solved ?

We should reduce impromptu EVA to the minimal, but still leave it as a possibility to do some emergency EVA.

Normally EVA must be done via the airlock schedule.

Is it a good idea ?

**Comment by bevans2000** [comment]
Yes we could use the schedule eventually pattern. But when the event time arrives what does it trigger? What does the Person do before the event if they are doing an Eva task?

**Comment by bevans2000** [comment]
Yes we could use the Group activity concept to support training. I just need to work out how a member is "rewarded" at the end.

As to the schedule event trigger, basically is just a callback method on the future so we can do anything against it. The key purpose is to stop polling the clock and make everything event driven

**Comment by mokun** [comment]
> work out how a member is "rewarded" at the end.

Not only should we add EVA operation skill points in their SkillManager, how about they need to take an exam and be score and be certified.

In the first orbit, everyone should be refreshed once every 3 months in order to perform EVA.

In the second orbit, we could extend that to every 6 months.

If they spend time reading up (using the Read class) EVA subject material, they may score better in the exam.
